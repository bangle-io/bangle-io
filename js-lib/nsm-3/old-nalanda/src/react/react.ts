import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
