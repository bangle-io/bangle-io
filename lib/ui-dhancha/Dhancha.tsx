import './style.css';

import type { ReactNode } from 'react';
import React, { useRef } from 'react';

import { vars } from '@bangle.io/css-vars';

import { useStickyNavigation } from './use-sticky-navigation';

/**
 * Provides the base structure of the app.
 * Does not handle scrolling and setting max height, so it is recommended to set that up in child component.
 * @returns
 */
export function Dhancha({
  widescreen = true,
  activitybar,
  workspaceSidebar,
  mainContent,
  noteSidebar,
}: {
  widescreen: boolean;
  activitybar: ReactNode;
  workspaceSidebar?: ReactNode;
  mainContent?: ReactNode;
  noteSidebar?: ReactNode;
}) {
  const activitybarRef = useRef<HTMLDivElement>(null);

  useStickyNavigation(widescreen, activitybarRef);

  const mainChild = (
    <main
      className={
        'B-ui-dhancha_main-content' + (widescreen ? ' BU_widescreen' : '')
      }
    >
      {mainContent}
    </main>
  );

  let childContent;

  if (widescreen) {
    childContent = (
      <>
        {workspaceSidebar && (
          <header
            style={{
              width: vars.misc.workspaceSidebarWidth,
            }}
            className="z-5 flex-shrink-0 h-screen border-r-1 border-colorNeutralBorder"
          >
            {workspaceSidebar}
          </header>
        )}
        {mainChild}
      </>
    );
  }
  // if mobile either show wSidebar or mainChild
  else {
    childContent = <>{workspaceSidebar ? workspaceSidebar : mainChild}</>;
  }

  return (
    <div
      className={
        'B-ui-dhancha_container' + (widescreen ? ' BU_widescreen' : '')
      }
    >
      <div
        role="navigation"
        aria-label="Activity Bar"
        ref={activitybarRef}
        className={
          'B-ui-dhancha_activitybar' + (widescreen ? ' BU_widescreen' : '')
        }
      >
        {activitybar}
      </div>

      {childContent}

      {widescreen && noteSidebar && (
        <aside
          className="B-ui-dhancha_note-sidebar shrink-0 h-screen z-5 border-l-1 border-colorNeutralBorder"
          style={{
            width: vars.misc.noteSidebarWidth,
          }}
        >
          {noteSidebar}
        </aside>
      )}
    </div>
  );
}

export function MultiColumnMainContent({ children }: { children: ReactNode }) {
  return (
    <div className="B-ui-dhancha_multi-column-main-content">{children}</div>
  );
}
