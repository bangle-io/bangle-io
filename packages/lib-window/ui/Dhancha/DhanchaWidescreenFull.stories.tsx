import type { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';

import { DhanchaWidescreen } from './index';

const meta: Meta<typeof DhanchaWidescreen> = {
  title: 'ui-dhancha/DhanchaWidescreenFull',
  component: DhanchaWidescreen,
  argTypes: {},
  decorators: [],
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof DhanchaWidescreen>;

export const Propper: Story = {
  render: () => {
    return <App />;
  },
};

function App() {
  const [showLeft, updateShowLeft] = useState();
  const [showRight, updateShowRight] = useState();

  return (
    <DhanchaWidescreen
      mainContent={<MainContent />}
      leftAside={<LeftSidebar />}
      rightAside={<NoteSidebar />}
      titlebar={
        <div className="bg-colorBgLayerFloat h-full border-b border-b-1 border-colorBorderSubdued">
          I am titlebar
        </div>
      }
    />
  );
}

function MainContent() {
  return (
    <div className="bg-colorBgLayer">
      <div className="bg-colorBgLayerFloat">I am main content</div>
    </div>
  );
}

function LeftSidebar() {
  return (
    <div
      className="bg-colorBgLayerFloat"
      style={{
        height: '100%',
        overflowY: 'scroll',
      }}
    >
      <div>Left bar</div>
    </div>
  );
}

function NoteSidebar() {
  return (
    <div
      style={{
        height: '100%',
        overflowY: 'scroll',
      }}
    >
      <div>Note sidebar</div>
    </div>
  );
}
