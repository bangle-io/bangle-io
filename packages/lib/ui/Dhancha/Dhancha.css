.B-ui-dhancha-widescreen {
  height: 100vh;
  display: grid;
  grid-template-rows: var(--BV-miscTitlebarHeight) calc(
      100vh - var(--BV-miscTitlebarHeight)
    );

  grid-template-columns: var(--BV-miscActivitybarWidth) auto;
  grid-template-areas:
    'activitybar titlebar-container'
    'activitybar main-container';
}

.B-ui-dhancha-widescreen.B-ui-has-left-aside {
  grid-template-columns:
    var(--BV-miscActivitybarWidth) var(--BV-miscLeftAsideWidth)
    auto;
  grid-template-areas:
    'activitybar left-aside titlebar-container'
    'activitybar left-aside main-container';
}

.B-ui-dhancha-widescreen.B-ui-has-right-aside {
  grid-template-columns: var(--BV-miscActivitybarWidth) auto var(
      --BV-miscRightAsideWidth
    );
  grid-template-areas:
    'activitybar titlebar-container titlebar-container'
    'activitybar main-container right-aside';
}

.B-ui-dhancha-widescreen.B-ui-has-right-aside.B-ui-has-left-aside {
  grid-template-columns:
    var(--BV-miscActivitybarWidth) var(--BV-miscLeftAsideWidth)
    auto var(--BV-miscRightAsideWidth);
  grid-template-areas:
    'activitybar left-aside titlebar-container titlebar-container'
    'activitybar left-aside main-container right-aside';
}

.BU_smallscreen .B-ui-dhancha-titlebar {
  display: flex;
  flex-direction: row;
  flex-shrink: 0;
  height: var(--BV-miscTitlebarHeight);
  position: sticky;
  border-bottom: 1px solid transparent;
  top: 0;
  width: 100%;
  z-index: 10;
}

/* the .up and .down are added in small screens */
.BU_smallscreen .B-ui-dhancha-titlebar.B-ui-dhancha_up {
  border-bottom-color: transparent;
  transition: border-bottom-color 0.3s ease-in;
}

.BU_smallscreen .B-ui-dhancha-titlebar.B-ui-dhancha_down {
  border-bottom-color: var(--BV-colorNeutralBorder);
}
