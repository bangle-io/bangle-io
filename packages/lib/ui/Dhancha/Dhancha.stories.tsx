import type { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';

import { DhanchaWidescreen } from './index';

const meta: Meta<typeof DhanchaWidescreen> = {
  title: 'ui-dhancha/DhanchaWidescreen',
  component: DhanchaWidescreen,
  argTypes: {},
  decorators: [],
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof DhanchaWidescreen>;

export const WithAll: Story = {
  args: {
    leftAside: (
      <div
        style={{
          height: '100%',
          overflowY: 'scroll',
        }}
      >
        <div>Activity bar</div>
        <WorkspaceSidebar />
      </div>
    ),
    mainContent: <MainContent />,
    rightAside: <NoteSidebar />,
    titlebar: <div className="bg-colorBgLayerFloat">I am titlebar</div>,
  },
};

export const OnlyRight: Story = {
  args: {
    mainContent: <MainContent />,
    rightAside: <NoteSidebar />,
    titlebar: <div className="bg-colorBgLayerFloat">I am titlebar</div>,
  },
};

export const OnlyLeft: Story = {
  args: {
    leftAside: (
      <div
        style={{
          height: '100%',
          overflowY: 'scroll',
        }}
      >
        <div>Activity bar</div>
        <WorkspaceSidebar />
      </div>
    ),
    mainContent: <MainContent />,
    titlebar: <div className="bg-colorBgLayerFloat">I am titlebar</div>,
  },
};

export const OnlyMain: Story = {
  args: {
    mainContent: <MainContent />,
    titlebar: <div className="bg-colorBgLayerFloat">I am titlebar</div>,
  },
};

function MainContent() {
  const [noteSidebarOpen, toggleNoteSidebar] = useState(true);
  const [workspaceSidebarOpen, toggleWorkspaceSidebar] = useState(true);

  return (
    <div
      style={{
        backgroundColor: 'pink',
        overflowY: 'scroll',
        height: '100%',
      }}
    >
      <div>
        Main content
        <button
          onClick={() => {
            toggleNoteSidebar((r) => !r);
          }}
        >
          Note Sidebar
        </button>
        <button
          onClick={() => {
            toggleWorkspaceSidebar((r) => !r);
          }}
        >
          Workspace Sidebar
        </button>
        <div>{generateText({ lines: 111 })}</div>
      </div>
    </div>
  );
}

function NoteSidebar() {
  const [noteSidebarOpen, toggleNoteSidebar] = useState(true);

  return noteSidebarOpen ? (
    <div
      style={{
        backgroundColor: 'lightblue',
        height: '100%',
        overflowY: 'scroll',
      }}
    >
      <div>Note sidebar</div>
      <div>{generateText({ lines: 10 })}</div>
    </div>
  ) : null;
}

function WorkspaceSidebar() {
  const [workspaceSidebarOpen, toggleWorkspaceSidebar] = useState(true);

  return workspaceSidebarOpen ? (
    <div
      style={{
        backgroundColor: 'lightgreen',
      }}
      onClick={() => {
        toggleWorkspaceSidebar((r) => !r);
      }}
    >
      <div>{generateText({ lines: 10 })}</div>
    </div>
  ) : null;
}

function generateText({ lines = 10 }) {
  return (
    <div>
      {Array.from({ length: lines }).map((_, i) => (
        <div key={i} style={{ marginTop: '1rem' }}>
          Tabbed UI Premise: make it clear which file I am working on and
          improve the top level organization of the app. I think it is time to
          just create a master UI skeleton that I can try out independently and
          progressively enhance it for the user. Again prone to the itch of
          creation, I need to think about what skeleton will work best for me.
          MemUI, logSeq and many more in my notes, have decent UI , I just need
          to capture it. Dhancha the main skeleton of the app. headerSection
          mainSection primaryColumn sidebar Notes nThis will be a good exercise
          to break up things in app directory, keep the in mind good article on
          css styling
          https://www.baldurbjarnason.com/2021/100-things-every-web-developer-should-know/
          also do some research on styling.
        </div>
      ))}
    </div>
  );
}
