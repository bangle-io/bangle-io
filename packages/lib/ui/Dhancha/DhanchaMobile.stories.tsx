import type { Meta, StoryObj } from '@storybook/react';
import React, { useState } from 'react';

import { DhanchaSmallscreen } from './index';

const meta: Meta<typeof DhanchaSmallscreen> = {
  title: 'ui-dhancha/DhanchaSmallscreen',
  component: DhanchaSmallscreen,
  argTypes: {},
  decorators: [],
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof DhanchaSmallscreen>;

export const Main: Story = {
  args: {
    titlebar: <div className="bg-colorBgLayerFloat w-full">I am titlebar</div>,
    mainContent: <MainContent />,
  },
};

function MainContent() {
  const [noteSidebarOpen, toggleNoteSidebar] = useState(true);
  const [workspaceSidebarOpen, toggleWorkspaceSidebar] = useState(true);

  return (
    <div
      style={{
        backgroundColor: 'pink',
        height: '100%',
      }}
    >
      <div>
        Main content
        <button
          onClick={() => {
            toggleNoteSidebar((r) => !r);
          }}
        >
          Note Sidebar
        </button>
        <button
          onClick={() => {
            toggleWorkspaceSidebar((r) => !r);
          }}
        >
          Workspace Sidebar
        </button>
        <div>{generateText({ lines: 111 })}</div>
      </div>
    </div>
  );
}

function NoteSidebar() {
  const [noteSidebarOpen, toggleNoteSidebar] = useState(true);

  return noteSidebarOpen ? (
    <div
      style={{
        backgroundColor: 'lightblue',
        height: '100vh',
        overflowY: 'scroll',
      }}
    >
      <div>Note sidebar</div>
      <div>{generateText({ lines: 10 })}</div>
    </div>
  ) : null;
}

function WorkspaceSidebar() {
  const [workspaceSidebarOpen, toggleWorkspaceSidebar] = useState(true);

  return workspaceSidebarOpen ? (
    <div
      style={{
        backgroundColor: 'lightgreen',
        height: '100vh',
        overflowY: 'scroll',
      }}
      onClick={() => {
        toggleWorkspaceSidebar((r) => !r);
      }}
    >
      <div>{generateText({ lines: 10 })}</div>
    </div>
  ) : null;
}

function generateText({ lines = 10 }) {
  return (
    <div>
      {Array.from({ length: lines }).map((_, i) => (
        <div key={i} style={{ marginTop: '1rem' }}>
          Tabbed UI Premise: make it clear which file I am working on and
          improve the top level organization of the app. I think it is time to
          just create a master UI skeleton that I can try out independently and
          progressively enhance it for the user. Again prone to the itch of
          creation, I need to think about what skeleton will work best for me.
          MemUI, logSeq and many more in my notes, have decent UI , I just need
          to capture it. Dhancha the main skeleton of the app. headerSection
          mainSection primaryColumn sidebar Notes nThis will be a good exercise
          to break up things in app directory, keep the in mind good article on
          css styling
          https://www.baldurbjarnason.com/2021/100-things-every-web-developer-should-know/
          also do some research on styling.
        </div>
      ))}
    </div>
  );
}
