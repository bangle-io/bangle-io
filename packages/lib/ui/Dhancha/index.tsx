import './Dhancha.css';

import type { ReactNode } from 'react';
import React, { useRef } from 'react';

import { vars } from '@bangle.io/css-vars';
import { cx } from '@bangle.io/utils';

import { useStickyNavigation } from './use-sticky-navigation';

/**
 * rightAside is expected to toggleable
 */
export function DhanchaWidescreen({
  activitybar,
  leftAside,
  mainContent,
  rightAside,
  titlebar,
}: {
  activitybar: ReactNode;
  leftAside?: ReactNode;
  mainContent: ReactNode;
  rightAside?: ReactNode;
  titlebar: ReactNode;
}) {
  const leftAsideStyle: React.CSSProperties = {
    gridArea: 'left-aside',
    // maxWidth: vars.misc.leftAsideWidth,
  };

  const titlebarContainerStyle: React.CSSProperties = {
    gridArea: 'titlebar-container',
  };

  const mainContentStyle: React.CSSProperties = {
    gridArea: 'main-container',
  };

  const rightAsideStyle: React.CSSProperties = {
    gridArea: 'right-aside',
    // maxWidth: vars.misc.rightAsideWidth,
  };

  return (
    <div
      className={cx(
        'B-ui-dhancha-widescreen',
        leftAside && 'B-ui-has-left-aside',
        rightAside && 'B-ui-has-right-aside',
      )}
    >
      <div
        role="navigation"
        aria-label="Title Bar"
        style={{
          gridArea: 'activitybar',
        }}
      >
        {activitybar}
      </div>
      {leftAside && (
        <aside style={leftAsideStyle} className="left-aside">
          {leftAside}
        </aside>
      )}
      <TitlebarContainer
        style={titlebarContainerStyle}
        className="titlebar-container"
      >
        {titlebar}
      </TitlebarContainer>
      <div style={mainContentStyle} className="main-content">
        {mainContent}
      </div>
      {rightAside && (
        <aside style={rightAsideStyle} className="right-aside">
          {rightAside}
        </aside>
      )}
    </div>
  );
}

export function DhanchaSmallscreen({
  titlebar,
  mainContent,
}: {
  titlebar: ReactNode;
  mainContent: ReactNode;
}) {
  const titlebarRef = useRef<HTMLDivElement>(null);

  useStickyNavigation(titlebarRef);

  return (
    <div>
      <TitlebarContainer ref={titlebarRef}>{titlebar}</TitlebarContainer>
      <div>{mainContent}</div>
    </div>
  );
}

const TitlebarContainer = React.forwardRef<
  HTMLDivElement,
  { children: ReactNode; style?: React.CSSProperties; className?: string }
>(function TitlebarContainer(props, ref) {
  return (
    <div
      role="navigation"
      aria-label="Title Bar"
      ref={ref}
      className={'B-ui-dhancha-titlebar ' + props.className}
      style={props.style}
    >
      {props.children}
    </div>
  );
});
