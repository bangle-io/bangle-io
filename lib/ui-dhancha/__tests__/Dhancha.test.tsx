/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { render } from '@testing-library/react';
import React from 'react';

import { MultiColumnMainContent } from '..';
import { Dhancha } from '../Dhancha';

test('renders', () => {
  let result = render(
    <div>
      <Dhancha
        widescreen={true}
        activitybar={<div>Activitybar</div>}
        mainContent={<div>Main content</div>}
        noteSidebar={<div>Note sidebar</div>}
        workspaceSidebar={<div>Workspace sidebar</div>}
      />
    </div>,
  );

  expect(result.container.innerHTML).toContain('Activitybar');
  expect(result.container.innerHTML).toContain('Note sidebar');
  expect(result.container.innerHTML).toContain('Main content');
  expect(result.container.innerHTML).toContain('Workspace sidebar');

  expect(result.container).toMatchSnapshot();
});

test('renders when widescreen=false and there is workspaceSidebar', () => {
  let result = render(
    <div>
      <Dhancha
        widescreen={false}
        activitybar={<div>Activitybar</div>}
        mainContent={<div>Main content</div>}
        noteSidebar={<div>Note sidebar</div>}
        workspaceSidebar={<div>Workspace sidebar</div>}
      />
    </div>,
  );

  expect(result.container.innerHTML).toContain('Activitybar');
  expect(result.container.innerHTML).not.toContain('Note sidebar');
  expect(result.container.innerHTML).not.toContain('Main content');
  expect(result.container.innerHTML).toContain('Workspace sidebar');

  expect(result.container).toMatchSnapshot();
});

test('renders when widescreen=false and there is no workspaceSidebar', () => {
  let result = render(
    <div>
      <Dhancha
        widescreen={false}
        activitybar={<div>Activitybar</div>}
        mainContent={<div>Main content</div>}
        noteSidebar={<div>Note sidebar</div>}
      />
    </div>,
  );

  expect(result.container.innerHTML).toContain('Activitybar');
  expect(result.container.innerHTML).not.toContain('Note sidebar');
  expect(result.container.innerHTML).toContain('Main content');

  expect(result.container).toMatchSnapshot();
});

describe('MultiColumnMainContent', () => {
  test('renders', () => {
    let result = render(
      <div>
        <MultiColumnMainContent>Hello</MultiColumnMainContent>
      </div>,
    );

    expect(result.container).toMatchSnapshot();
  });
});
