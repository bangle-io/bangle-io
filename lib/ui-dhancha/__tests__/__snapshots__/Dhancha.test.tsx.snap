// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiColumnMainContent renders 1`] = `
<div>
  <div>
    <div
      class="B-ui-dhancha_multi-column-main-content"
    >
      Hello
    </div>
  </div>
</div>
`;

exports[`renders 1`] = `
<div>
  <div>
    <div
      class="B-ui-dhancha_container BU_widescreen"
    >
      <div
        aria-label="Activity Bar"
        class="B-ui-dhancha_activitybar BU_widescreen"
        role="navigation"
      >
        <div>
          Activitybar
        </div>
      </div>
      <header
        class="z-5 flex-shrink-0 h-screen border-r-1 border-colorNeutralBorder"
      >
        <div>
          Workspace sidebar
        </div>
      </header>
      <main
        class="B-ui-dhancha_main-content BU_widescreen"
      >
        <div>
          Main content
        </div>
      </main>
      <aside
        class="B-ui-dhancha_note-sidebar shrink-0 h-screen z-5 border-l-1 border-colorNeutralBorder"
      >
        <div>
          Note sidebar
        </div>
      </aside>
    </div>
  </div>
</div>
`;

exports[`renders when widescreen=false and there is no workspaceSidebar 1`] = `
<div>
  <div>
    <div
      class="B-ui-dhancha_container"
    >
      <div
        aria-label="Activity Bar"
        class="B-ui-dhancha_activitybar"
        role="navigation"
      >
        <div>
          Activitybar
        </div>
      </div>
      <main
        class="B-ui-dhancha_main-content"
      >
        <div>
          Main content
        </div>
      </main>
    </div>
  </div>
</div>
`;

exports[`renders when widescreen=false and there is workspaceSidebar 1`] = `
<div>
  <div>
    <div
      class="B-ui-dhancha_container"
    >
      <div
        aria-label="Activity Bar"
        class="B-ui-dhancha_activitybar"
        role="navigation"
      >
        <div>
          Activitybar
        </div>
      </div>
      <div>
        Workspace sidebar
      </div>
    </div>
  </div>
</div>
`;
